### DISCUSSION TITLE
Streams not buffering enough - current situation

### DISCUSSION THREAD

**Discussion by dark-penguin** [smplayer-dev/smplayer#256] [Q&A]
I'm investigating what can I do about this problem: yt-dlp downloads a video in seconds, but opening the same link in smplayer buffers only a few seconds ahead. I want it to download at least half an hour in advance, because I like to "skim" videos very quickly.

Additional mpv options I'm using are: 
`--profile=gpu-hq --gpu-context=x11egl --video-sync=display-resample --interpolation --tscale=oversample --cache=yes --cache-secs=1800 --demuxer-seekable-cache=yes --demuxer-max-bytes=512MiB --demuxer-readahead-secs=1800 --ytdl-format=137+140/299+140/22/136+140/bestvideo[height<=?1200]+bestaudio/best`
So, the problem is not the buffer being too small. (Or am I using the wrong options?..)

What I've found so far is:
- yt-dlp: this has nothing to do with yt-dlp, because it's only used to get the URLs, not downloading
- Smplayer: this has nothing to do with smplayer, because those parameters to mpv
- MPV: this has nothing to do with mpv, because mpv just passes the URL to ffmpeg
- FFmpeg: does not support external downloaders, but it would help if it could use aria2c with specific parameters

Sources:
https://github.com/mpv-player/mpv/issues/9149
https://github.com/mpv-player/mpv/issues/8655

However, since moving from youtube-dl to yt-dlp helped with throttling, maybe I'm misunderstanding something or being not up to date. So I want to confirm, is this the correct overall situation at the moment? Maybe something was already fixed in the latest versions on some of those levels, or maybe there is now a way to hack this with scripts or parameters to FFmpeg?

EDIT: the latest version of smplayer (as suggested in #247 ) helps with the "constant buffering" problem by introducing obvious yt-dlp support, but does not help with buffering more than a few seconds ahead. This makes me think that if you can fix buffering by replacing youtube-dl with yt-dlp, then maybe you can fix "more buffering" as well on some level other than FFmpeg?..

**Comment by KilliK69** [comment]
Google started throttling the downloading of Youtube videos from external apps like youtube-dl or Jdownloader a few weeks back.
also the maintainer of the youtube-dl has been inactive since June, and the project is in an abandonment state right now.
the good news is it was forked to the youtube-dlp project, which is in active development and addresses the throttling problem from google.

the current stable version of SMplayer still uses the old youtube-dl version, which causes the lagging when playing youtube videos.
but the latest development version comes with the youtube-dlp version which fixes the problem. you can download it from here:
https://github.com/smplayer-dev/smplayer/releases/tag/v21.8.0.9999
after you install it, go to preferences -> network and select yt-dlp instead of youtube-dl as youtube application. 

TLDR:
use youtube-dlp to save videos from the youtube site.
use the latest development version of SMplayer to play youtube videos directly from the site.

**Reply by dark-penguin** [reply]
I'm aware that youtube-dl no longer works. I've already switched to yt-dlp, and symlinked it to "youtube-dl" so that smplayer 20.6.0 from the Debian Bullseye repos can find it. The issue I'm talking about is a little different.

Moving to yt-dlp fixed the problem of videos being completely unplayable, always "buffering". But it's still downloading ahead very slowly, which I can see by my network activity monitor. It's enough to play the video with no problem, but not enough to scroll quickly, unless I let it sit on pause for a very long time. And if I directly use yt-dlp to "download" the whole video, then it gets done within seconds.

According to what I've found on the Internet, seems like this can only be fixed on the ffmpeg layer, because whatever magic yt-dlp is employing is not used in smplayer/mpv. But since switching to yt-dlp somehow fixed it (even though it shouldn't have been possible), maybe there's something else I can do to further improve the buffering speed?..

I don't understand how did switching to yt-dlp help, if it's only used to extract the URL. Is it not only used to extract the URL, or did youtube-dl simply return wrong URLs?..
